### PR TITLE
DATETIME handling & microsecond support

### PR DESCRIPTION
**Pull request by Hydrocharged**

### CODE REVIEW

File: sql/type.go [hunk 1]
```diff
@@ -700,7 +700,19 @@ func (t datetimeT) Zero() interface{} {
...
-const DatetimeLayout = "2006-01-02 15:04:05"
+const DatetimeLayout = "2006-01-02 15:04:05.999999"
+
+// DatetimeLayouts hold extra configurations allowed for parsing. It does
+// not have all the layouts supported by mysql. Missing are two digit year
+// versions of common cases and dates that use non common separators.
+//
+// https://github.com/MariaDB/server/blob/mysql-5.5.36/sql-common/my_time.c#L124
+var DatetimeLayouts = []string{
+	"2006-01-02",
```
Comment (zachmu) [inline sql/type.go]: I would include the main constant in this slice, then just loop over it a single time

File: sql/type.go [hunk 2]
```diff
@@ -731,7 +743,18 @@ func (t datetimeT) Convert(v interface{}) (interface{}, error) {
 	case string:
 		t, err := time.Parse(DatetimeLayout, value)
```
Comment (zachmu) [inline sql/type.go]: This is a little convoluted. Better to loop over the accepted layouts a single time, returning as soon as you find one that parses.

File: sql/type.go [hunk 3]
```diff
@@ -731,7 +743,18 @@ func (t datetimeT) Convert(v interface{}) (interface{}, error) {
...
+			for _, fmt := range DatetimeLayouts {
+				if t2, err2 := time.Parse(fmt, value); err2 == nil {
+					t = t2
+					failed = false
+					break
+				}
+			}
+
+			if failed {
+				return nil, ErrConvertingToTime.Wrap(err, v)
```
Comment (zachmu) [inline sql/type.go]: This wrapped error is no longer applicable (it will give the false impression that the last format attempted is the only correct one)

**Comment by Hydrocharged** [comment]
I copied the logic from the Timestamps section lol. I’ll add these adjustments there as well.

**Pull request merged**